### PR TITLE
Refactor the initialization of the editor, require only a post ID

### PR DESCRIPTION
**Pull request by youknowriad**
Previously to initialize the Gutenberg `edit-post` module we had to pass a post object to the `initializeEditor` function. This PR updates this behavior and requires only a `postId`. A `defaultPost` is still kept to allow passing intialization attributes (essentially for demo content)

There are several reasons for this:

 - Making the editor more independent from the backend
 - Leveraging the data module to fetch the post objects
 - And more importantly, this prepares the way to move post saving into `edit-post` (relying on the `data` module).
 - It also simplifies a lot of code in `client-assets` and thanks to the preloading support in `wp.apiRequest`, the post is still loaded instantaneously.

**Testing instructions**

 - Test that you can create a new post
 - you can open an existing post
 - you can load the demo content.

### CODE REVIEW

File: lib/client-assets.php [hunk 1]
```diff
@@ -966,7 +895,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	$script .= <<<JS
 		window._wpLoadGutenbergEditor = wp.api.init().then( function() {
 			wp.blocks.registerCoreBlocks();
-			return wp.editPost.initializeEditor( 'editor', window._wpGutenbergPost, editorSettings );
+			return wp.editPost.initializeEditor( 'editor', window._wpGutenbergPostId, editorSettings, window._wpGutenbergDefaultPost );
```
Comment (gziolo) [inline lib/client-assets.php]: Can you simply inline `_wpGutenbergPostId` and `_wpGutenbergDefaultPost`?

Comment (youknowriad) [reply lib/client-assets.php]: Not sure what you mean? The reason those are different and not at the same position is that one is mandatory while the other is optional and could probably be removed later if we remove the demo content.

Comment (gziolo) [reply lib/client-assets.php]: Why not wrap this code with `sprintf` and inject those 2 variables directly in here instead of having globals?

Comment (gziolo) [reply lib/client-assets.php]: It's difficult to express with PHP but simplified version would be:
```php
$init_script = <<<JS
    ...
    return wp.editPost.initializeEditor( 'editor', %s, editorSettings, %s );
    ...
JS;
$script .= sprintf( init_script, $post->ID, wp_json_encode( array( 'title' => ... ) ) );
````

Comment (youknowriad) [reply lib/client-assets.php]: Gotcha, `window._wpGutenbergDefaultPost` can be undefined in some cases but I guess it's feasible.

Comment (youknowriad) [reply lib/client-assets.php]: Ok, this doesn't work easily because of the content of the demo post is a JavaScript file. We could try to convert it to JSON but seems not worth it.

File: edit-post/index.js [hunk 2]
```diff
@@ -33,15 +33,17 @@ window.jQuery( document ).on( 'heartbeat-tick', ( event, response ) => {
  * an unhandled error occurs, replacing previously mounted editor element using
  * an initial state from prior to the crash.
  *
+ * @param {Object}  postId   ID of the post to edit.
  * @param {Element} target   DOM node in which editor is rendered.
  * @param {?Object} settings Editor settings object.
+ * @param {Object}  defaultPost   Post initilization object
```
Comment (aduth) [inline edit-post/index.js]: Nit: Alignment of parameters.

File: lib/client-assets.php [hunk 3]
```diff
@@ -844,31 +777,20 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
...
-		);
-	}
+	$is_new_post = 'auto-draft' === $post->post_status;
 
 	// Set the post type name.
 	$post_type = get_post_type( $post );
 
 	// Preload common data.
 	$preload_paths = array(
+		sprintf( '/wp/v2/posts/%s?context=edit', $post->ID ),
```
Comment (aduth) [inline lib/client-assets.php]: Nit: Placeholder could be `%d`, expecting numeric.

File: lib/client-assets.php [hunk 4]
```diff
@@ -829,31 +762,20 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
...
 		sprintf( '/wp/v2/types/%s?context=edit', $post_type ),
 		sprintf( '/wp/v2/users/me?post_type=%s&context=edit', $post_type ),
 		'/wp/v2/taxonomies?context=edit',
-		gutenberg_get_rest_link( $post_to_edit, 'about', 'edit' ),
```
[the remark this answers is not in the document]
Comment (aduth) [reply lib/client-assets.php]: >Not certain what this call was preloading but I didn't notice any regression in term of preloaded requests, let me know if I'm missing something.

I guess it was preloading the post type data. Unclear where that was used, but I cannot see anything immediately that should break. Probably safe. I probably should have documented it better in the first place 😬

File: lib/client-assets.php [hunk 5]
```diff
@@ -883,10 +805,9 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		'before'
 	);
 
-	// Initialize the post data.
 	wp_add_inline_script(
 		'wp-edit-post',
-		'window._wpGutenbergPost = ' . wp_json_encode( $post_to_edit ) . ';'
+		sprintf( 'window._wpGutenbergPostId = %s;', $post->ID )
```
Comment (aduth) [inline lib/client-assets.php]: Should we drop the Gutenberg codename ? 😄

File: lib/client-assets.php [hunk 6]
```diff
@@ -991,8 +920,9 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	 * Scripts
 	 */
 	wp_enqueue_media( array(
-		'post' => $post_to_edit['id'],
-	) );
+		'post' => $post->ID,
+
+		) );
```
Comment (aduth) [inline lib/client-assets.php]: Odd tabbing.

File: edit-post/editor.js [hunk 7]
```diff
@@ -9,16 +9,26 @@ import { EditorProvider, ErrorBoundary } from '@wordpress/editor';
...
+	}
+
+	const editorSettings = {
+		...settings,
+		hasFixedToolbar,
+	};
+
 	return (
-		<EditorProvider settings={ { ...settings, hasFixedToolbar } } { ...props }>
+		<EditorProvider settings={ editorSettings } post={ { ...post, ...defaultPost } } { ...props }>
```
Comment (aduth) [inline edit-post/editor.js]: Shouldn't the spread be the other way around? `{ ...defaultPost, ...post }`

**Comment by youknowriad** [comment]
I updated the PR and added dynamic loading of "entities" given a "kind" in f776128

Before, I go further by adding unit tests etc... I'd appreciate some thoughts on the approach here and ideas to improve it.

The idea is the following: 

- We need to automatically load a post given its id and post type using `getEntityRecord( "postType", "post", id )` 

 - And this should work for any post or Custom Post Type which means we can't statically load the posts from `/wp/v2/posts`, we need to map each post type to its rest base.

 - So what I did in the resolver `getEntityRecord` I added a mechanism to load the entities associated with a kind (in our example the kind is `postType`). So what happens is if you call `getEntityRecord` if the requested entity is not loaded we try to dynamically load the config for all the entities of the given kind.

Thoughs @gziolo @aduth @WordPress/gutenberg-core

File: core-data/entities.js [hunk 8]
```diff
@@ -1,23 +1,36 @@
...
- * @param {string} kind  Entity kind.
- * @param {string} name  Entity name.
- *
- * @return {Object} Entity
+ * @return {Array} entities
  */
-export function getEntity( kind, name ) {
-	return find( entities, { kind, name } );
+async function loadPostTypeEntities() {
+	const postTypes = await apiRequest( { path: '/wp/v2/types?context=edit' } );
```
Comment (aduth) [inline core-data/entities.js]: Is `?context=edit` valid for the types endpoint? I'm seeing an error when trying to request it directly this way.

File: core-data/resolvers.js [hunk 9]
```diff
@@ -33,16 +43,37 @@ export async function* getAuthors() {
...
  *
  * @param {Object} state  State tree
  * @param {string} kind   Entity kind.
  * @param {string} name   Entity name.
  * @param {number} key    Record's key
  */
 export async function* getEntityRecord( state, kind, name, key ) {
-	const entity = getEntity( kind, name );
+	yield* loadKindEntities( state, kind );
+	// I can't use the state because it's outdated at this point
```
Comment (aduth) [inline core-data/resolvers.js]: Could we use the yielded result from `loadKindEntities`?

File: core-data/resolvers.js [hunk 10]
```diff
@@ -33,16 +43,37 @@ export async function* getAuthors() {
 	yield receiveUserQuery( 'authors', users );
 }
 
+async function* loadKindEntities( state, kind ) {
+	const hasEntities = hasEntitiesByKind( state, kind );
```
Comment (aduth) [inline core-data/resolvers.js]: What is this doing?

File: core-data/selectors.js [hunk 11]
```diff
@@ -76,6 +76,31 @@ export function getUserQueryResults( state, queryID ) {
...
+/**
+ * Returns whether the entities for the give kind are loaded.
+ *
+ * @param {Object} state   Data state.
+ * @param {string} kind  Entity kind.
+ *
+ * @return {boolean} Whether the entities are loaded
+ */
+export function hasEntitiesByKind( state, kind ) {
+	return state.entities.config[ kind ] !== undefined;
```
Comment (aduth) [inline core-data/selectors.js]: `entities.config` is an array? When will this ever not be `false` ?

File: core-data/reducer.js [hunk 12]
```diff
@@ -129,23 +129,85 @@ function entity( entityConfig ) {
...
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export const entities = ( state = {}, action ) => {
+	const newConfig = entitiesConfig( state.config, action );
+
+	// Generates a dynamic reducer for the entities
+	const entitiesByKind = groupBy( newConfig, 'kind' );
```
Comment (aduth) [inline core-data/reducer.js]: Does this need to be run on every dispatch? Or could we limit it to those which would actually impact the configuration to justify dynamically creating new reducers?

File: core-data/reducer.js [hunk 13]
```diff
@@ -129,23 +129,85 @@ function entity( entityConfig ) {
...
+
+/*
+ * Reducer managing posts state. Keyed by ID.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export function posts( state = {}, action ) {
```
Comment (aduth) [inline core-data/reducer.js]: Do we want this reducer? Isn't this what entities is meant to be providing on our behalf?

File: core-data/resolvers.js [hunk 14]
```diff
@@ -33,16 +43,37 @@ export async function* getAuthors() {
 	yield receiveUserQuery( 'authors', users );
 }
 
+async function* loadKindEntities( state, kind ) {
```
Comment (aduth) [inline core-data/resolvers.js]: This isn't so much a resolver, is it? More a utility?

**Review by aduth — COMMENTED** [review]
The fact we have `loadKindEntities` as a utility within `resolvers.js` has me wondering if we should have it be a proper resolver; i.e. have a reducer which returns the entity configuration for a given post type, and resolve it if necessary, so it's not treated any different than another selector. Would that be possible?

File: core-data/reducer.js (hunk 12 above)
Comment (youknowriad) [reply core-data/reducer.js]: Do you I should use a local variable to keep track of the current reducer?

File: core-data/resolvers.js (hunk 14 above)
Comment (youknowriad) [reply core-data/resolvers.js]: Well, the difference is not totally clear here since it uses selectors and yield actions.
Anyway, I moved it to the `entities.js` file for now.

File: core-data/resolvers.js (hunk 9 above)
Comment (youknowriad) [reply core-data/resolvers.js]: The problem is `loadKindEntities` don't do anything if the entities are loaded. It can be updated to return the entities in all cases. So it can be renamed `getKindEntities` which could be just a regular resolver for the `getKindEntities` selector and that's the point where I felt the need of a `resolve` API.

`yield resolve( 'core' ).getKindEntities()` or something like that where I won't be obliged to call the selector `getKindEntities` by myself inside the resolver but it will be called automatically when the resolver finishes. 

I guess it's still not clear if it will be a common use-case etc... I prefer to wait a bit with and keep the current "shortcut"

File: core-data/entities.js (hunk 8 above)
Comment (youknowriad) [reply core-data/entities.js]: Weird, It worked for me (I removed the preloading and I do see the request in the network tab)

**Comment by youknowriad** [comment]
Let me know if you'll agree with the ideas here so I can move forward with the testing etc...

Also, I'm looking forward using the data module to extract saving posts from the `editor` module into the `edit-post` module.

File: core-data/resolvers.js (hunk 9 above)
Comment (youknowriad) [reply core-data/resolvers.js]: Ok so to simplify unit-testing, I went ahead and returned the available entities in all use-cases.

File: core-data/reducer.js (hunk 12 above)
Comment (aduth) [reply core-data/reducer.js]: Maybe it's included as part of the `state` value?

File: edit-post/editor.js (hunk 7 above)
Comment (aduth) [reply edit-post/editor.js]: >Shouldn't the spread be the other way around? { ...defaultPost, ...post }

Still wondering about this note.

Comment (aduth) [reply edit-post/editor.js]: Okay, I see this may be needed because auto-draft title replacement is meant to take precedent over the default title. I wonder if we could...

- Avoid needing this override in the first place (set auto-draft title as it's created?)
- Rename the variable to reflect that it's an override, not a set of fallback values
- Comment to reflect the intended usage

File: core-data/entities.js [hunk 15]
```diff
@@ -34,4 +53,28 @@ export const getMethodName = ( kind, name ) => {
...
+
+	const kindConfig = find( kinds, { name: kind } );
+	if ( ! kindConfig ) {
+		return [];
+	}
+
+	entities = await kindConfig.loadEntities();
+	yield addEntities( entities );
+
+	return entities;
```
Comment (aduth) [inline core-data/entities.js]: The multipurposing of this function feels a bit awkward / forced. Not sure I have a great solution here. One option is to make it always yield the action, and extract this from the `getEntityRecord` resolver:

```js
export async function* getKindEntities( state, kind ) {
	const kindConfig = find( kinds, { name: kind } );
	if ( ! kindConfig ) {
		return;
	}

	let entities = getEntitiesByKind( state, kind );
	if ( ! entities ) {
		entities = await kindConfig.loadEntities();
	}

	yield addEntities( entities );
}
```

```js
export async function* getEntityRecord( state, kind, name, key ) {
	const { entities } = await ( await getKindEntities( state, kind ).next() ).value;
	const entity = find( entities, { kind, name } );
	if ( ! entity ) {
		return;
	}
	const record = await apiRequest( { path: `${ entity.baseUrl }/${ key }?context=edit` } );
	yield receiveEntityRecords( kind, name, record );
}
```

Makes `getEntityRecord` have strong awareness of what's yielded by `getKindEntities`, which could break down if ever `getKindEntities` yields more or different than just the action of `ADD_ENTITIES`. We could make `getKindEntities` not a generator, which could simplify it a bit more.

In retrospect, I might be more open to the idea of using `select` from within the resolver as was originally proposed, if it avoids some of the complexity here.

File: edit-post/editor.js (hunk 7 above)
Comment (aduth) [reply edit-post/editor.js]: I think it's also needed for the new post, since auto-draft is created with a "(Auto-draft)" title and we don't want to show this in the editor, so we needed to be able to override, though I think ideally we just... don't assign that title 😄

File: core-data/reducer.js (hunk 12 above)
Comment (aduth) [reply core-data/reducer.js]: It does seem strange, though accurate if we're saying that the reducer itself is dependent on state, that it is itself state.

Or, if it's a derivation, then a selector may be appropriate? 

Entering unfamiliar territory for me with this dynamic reducer 😄

File: core-data/entities.js (hunk 15 above)
Comment (youknowriad) [reply core-data/entities.js]: I do agree, it's a bit weird, I'd like us to avoid adding complexity for the moment with a new API or something, let's try it for some time. We'll see how it plays when adding other entity selectors. I'm kind of hesitant to introduce `select` within a resolver just for the entities use-case.

**Comment by youknowriad** [comment]
Rebased this, it's ready for a final review

File: core-data/actions.js [hunk 16]
```diff
@@ -81,3 +95,4 @@ export function receiveThemeSupportsFromIndex( index ) {
 		themeSupports: index.theme_supports,
 	};
 }
+
```
Comment (aduth) [inline core-data/actions.js]: Nit: Unnecessary newline.

File: core-data/entities.js [hunk 17]
```diff
@@ -1,23 +1,42 @@
...
+
 /**
- * Returns the entity object given its kind and name.
- *
- * @param {string} kind  Entity kind.
- * @param {string} name  Entity name.
+ * Returns the list of post type entities.
  *
- * @return {Object} Entity
+ * @return {Array} entities
```
Comment (aduth) [inline core-data/entities.js]: Pedantry: An `async` function always returns type `Promise`†. In this case it's a promise which resolves with an array.

_† ...which ESLint makes quite annoying for its `valid-jsdoc` rule since it assumes the presence of a `return` statement, which isn't always the case:_

```js
async function sleep( duration ) {
	return new Promise( ( resolve ) => {
		setTimeout( resolve, duration );
	} );
}

async function resolve() {
	await sleep( 1000 );
	await sleep( 1000 );
}

console.log( resolve() instanceof Promise );
// true
```

File: core-data/resolvers.js [hunk 18]
```diff
@@ -33,15 +38,19 @@ export async function* getAuthors() {
...
+ * Requests an entity's record from the REST API.
  *
  * @param {Object} state  State tree
  * @param {string} kind   Entity kind.
  * @param {string} name   Entity name.
  * @param {number} key    Record's key
  */
 export async function* getEntityRecord( state, kind, name, key ) {
-	const entity = getEntity( kind, name );
+	const entities = yield* getKindEntities( state, kind );
```
Comment (aduth) [inline core-data/resolvers.js]: I've reached the extent of my knowledge of asynchronous generators 😄 Is an `await` keyword needed on this line if `getKindEntities` may only return (resolve) after an async `apiRequest` ?

File: edit-post/index.js [hunk 19]
```diff
@@ -47,13 +50,15 @@ export function reinitializeEditor( target, settings ) {
...
+ * @param {string}  id            Unique identifier for editor instance.
+ * @param {Object}  postId        ID of the post to edit.
+ * @param {Object}  postType      Post type of the post to edit.
+ * @param {?Object} settings      Editor settings object.
+ * @param {Object}  overridePost  Post properties to override.
  *
  * @return {Object} Editor interface.
  */
-export function initializeEditor( id, post, settings ) {
+export function initializeEditor( id, postId, postType, settings, overridePost ) {
```
Comment (aduth) [inline edit-post/index.js]: In a potential future where `postId` becomes optional, would we want to support calling this as simply...

```js
initializeEditor( 'gutenberg', 'post' );
```

i.e. with `postType` before the optional `postId`

Asides:

- `id` is pretty useless now. It was originally intended for separate instances of an editor, which we don't really support. Maybe it should just be dropped.
- At what point might we consider an object argument? Since we could have many variations (post with ID and no overrides, post with no ID and overrides, post with overrides but no settings)

**Review by aduth — APPROVED** [review]
The entities implementation is becoming a bit hard to follow in places, while simultaneously impressive in its coordinated result. I think this is an important refactoring, so would be inclined to move forward on it and iterate as necessary on entities in future iterations. Nice work 👍 

(Needs a rebase)

File: edit-post/index.js (hunk 19 above)
Comment (youknowriad) [reply edit-post/index.js]: I'll keep the `id` for now. I'm not certain if we'll support multi-instances at some point but prefer to keep it for now.

An object argument: I don't know 🤷‍♂️, It may feel like a component though and maybe keeping it as clarify the distinction?

**Comment by youknowriad** [comment]
I rebase this PR and tests are failing. The problem is that the REST API hooks adding `action-publish` flag are not being used when preloading data. I believe this is a "hooks" order issue. @danielbachhuber Any idea how to fix that? What hook should I use instead of `registered_post_type`? Any other idea?

**Comment by danielbachhuber** [comment]
> The problem is that the REST API hooks adding `action-publish` flag are not being used when preloading data.

Are you sure about this? I'm not able to reproduce the e2e test failure locally.

This doesn't seem like a load order issue to me. `registered_post_type` is much earlier than `admin_enqueue_scripts`.

**Comment by youknowriad** [comment]
@danielbachhuber In my testing yesterday the post object in the preloaded data didn't have the "action-publish". I'll try again in a bit to confirm.
